backend/models/loginRecord.js

```diff
@@ -15,113 +15,97 @@
-                version: userAgentObject.browser.version,
-                type: "browser"        
-        }
-        
-        userAgentInfo.engine = {
-                name: userAgentObject.engine.name,
-                version: userAgentObject.engine.version
-        }
-
-        userAgentInfo.os = {
-                osName: userAgentObject.os.name,
-                osVersion: userAgentObject.os.version
-        }
-
-        userAgentInfo.device = device(userAgentString).type;
-
-        return userAgentInfo;
-}
-
-async function getLocationFromIPAddress(ipAddress){
-        return await ip2location.fetch(ipAddress);
-}
-
-function isUserAgentFromPlugin(userAgent){        
-        return userAgent.split(" ")[0] == "PLUGIN:";
-}
-
- module.exports = LoginRecord;
\ No newline at end of file
+"use strict";
+const db = require("../handler/db");
+
+// detects edge as browser but not device
+const UaParserJs = require("ua-parser-js");
+const Device = require("device");
+const Ip2location = require("ip-to-location");
+
+const LoginRecord = {};
+
+LoginRecord.saveLoginRecord = async (req) => {
+	const loginRecord = {
+		_id: req.sessionID,
+		loginTime: new Date(),
+		ipAddr: req.ips[0] || req.ip
+	};
+
+	const referrer = req.header("Referer");
+	if (referrer) {
+		loginRecord.referrer = referrer;
+	}
+
+	const { country_name, city } = await getLocationFromIPAddress(loginRecord.ipAddr);
+	loginRecord.location = {
+		country: country_name,
+		city
+	};
+
+	const userAgentString = req.headers["user-agent"];
+	const uaInfo = isUserAgentFromPlugin(userAgentString) ?
+		getUserAgentInfoFromPlugin(userAgentString) : getUserAgentInfoFromBrowser(userAgentString);
+
+	loginRecord.application = uaInfo.application;
+	loginRecord.engine = uaInfo.engine;
+	loginRecord.os = uaInfo.os;
+	loginRecord.device = uaInfo.device;
+
+	await db.insert("loginRecords", req.body.username, loginRecord);
+
+};
+
+// Format:
+// PLUGIN: {OS Name}/{OS Version} {Host Software Name}/{Host Software Version} {Plugin Type}/{Plugin Version}
+// Example:
+// PLUGIN: Windows/10.0.19042.0 REVIT/2021.1 PUBLISH/4.15.0
+const getUserAgentInfoFromPlugin = (userAgentString) => {
+	const [osInfo, appInfo, engineInfo] = userAgentString.replace("PLUGIN: ", "").split(" ");
+
+	const osInfoComponents = osInfo.split("/");
+	const appInfoComponents = appInfo.split("/");
+	const engineInfoComponents = engineInfo.split("/");
+
+	const userAgentInfo = {
+		application: {
+			name: appInfoComponents[0],
+			version: appInfoComponents[1],
+			type: "plugin"
+		},
+		engine: {
+			name: "3drepoplugin",
+			version: engineInfoComponents[1]
+		},
+		os: {
+			name: osInfoComponents[0],
+			version: osInfoComponents[1]
+		},
+		device: "desktop"
+	};
+
+	return userAgentInfo;
+};
+
+const getUserAgentInfoFromBrowser = (userAgentString) => {
+
+	const { browser, engine, os } = UaParserJs(userAgentString);
+	const userAgentInfo = {
+		application: { ...browser, type: browser === null ? null : "browser"},
```

Comment:
doesn't look like this happens, i just tried it with `undefined` and an empty string, these are the results:

![image](https://user-images.githubusercontent.com/11945337/122401262-58542e80-cf74-11eb-8535-fe2468f58842.png)


So the library always return the object, just the names might be undefined.

so I think what we want to do here is this:
```suggestion
		application: browser.name ? {...browser, type: "browser"} : {type: "unknown"},
```